CommonTools/RecoAlgos/interface/FKDTree.h

```diff
@@ -0,0 +1,377 @@
+
+
+namespace{
+    const std::array<unsigned int,32> MultiplyDeBruijnBitPosition{{0, 9, 1, 10, 13, 21, 2, 29, 11, 14, 16, 18, 22, 25, 3, 30,
+  8, 12, 20, 28, 15, 17, 24, 7, 19, 27, 23, 6, 26, 5, 4, 31}};
+      unsigned int ilog2 (unsigned int v)
+    {
+    v |= v >> 1; // first round down to one less than a power of 2 
+    v |= v >> 2;
+    v |= v >> 4;
+    v |= v >> 8;
+    v |= v >> 16;
+    return MultiplyDeBruijnBitPosition[(unsigned int)(v * 0x07C4ACDDU) >> 27];
+
+    }
+
+
+/*    unsigned int ilog2 (unsigned int v)
+    {
+       return (unsigned int)floor(std::log2(v));
+    
+    }
+    */
+}
+
+
+
+
+template<class TYPE, unsigned int numberOfDimensions>
+class FKDTree
+{
+
+public:
+
+	FKDTree(const unsigned int nPoints)
+	{
+		theNumberOfPoints = nPoints;
+
+		theDepth = ilog2(nPoints);
+		for (auto& x : theDimensions)
+			x.resize(theNumberOfPoints);
+		theIntervalLength.resize(theNumberOfPoints, 0);
+		theIntervalMin.resize(theNumberOfPoints, 0);
+		theIds.resize(theNumberOfPoints);
+		thePoints.reserve(theNumberOfPoints);
+
+	}
+
+
+
+
+	FKDTree()
+	{
+		theNumberOfPoints = 0;
+		theDepth = 0;
+		for (auto& x : theDimensions)
+			x.clear();
+		theIntervalLength.clear();
+		theIntervalMin.clear();
+		theIds.clear();
+		thePoints.clear();
+	}
+
+	FKDTree(const FKDTree<TYPE, numberOfDimensions>& v);
+
+	FKDTree(FKDTree<TYPE, numberOfDimensions> && other)
+	{
+		theNumberOfPoints(std::move(other.theNumberOfPoints));
+		theDepth(std::move(other.theDepth));
+
+		theIntervalLength.clear();
+		theIntervalMin.clear();
+		theIds.clear();
+		thePoints.clear();
+		for (auto& x : theDimensions)
+			x.clear();
+
+		theIntervalLength = std::move(other.theIntervalLength);
+		theIntervalMin = std::move(other.theIntervalMin);
+		theIds = std::move(other.theIds);
+
+		thePoints = std::move(other.thePoints);
+		for (unsigned int i = 0; i < numberOfDimensions; ++i)
+			theDimensions = std::move(other.theDimensions);
+	}
+
+	FKDTree<TYPE, numberOfDimensions>& operator=(
+			FKDTree<TYPE, numberOfDimensions> && other)
+	{
+
+		if (this != &other)
+		{
+			theNumberOfPoints(std::move(other.theNumberOfPoints));
+			theDepth(std::move(other.theDepth));
+
+			theIntervalLength.clear();
+			theIntervalMin.clear();
+			theIds.clear();
+			thePoints.clear();
+			for (auto& x : theDimensions)
+				x.clear();
+
+			theIntervalLength = std::move(other.theIntervalLength);
+			theIntervalMin = std::move(other.theIntervalMin);
+			theIds = std::move(other.theIds);
+
+			thePoints = std::move(other.thePoints);
+			for (unsigned int i = 0; i < numberOfDimensions; ++i)
+				theDimensions = std::move(other.theDimensions);
+		}
+		return *this;
+
+	}
+
+	void push_back(const FKDPoint<TYPE, numberOfDimensions>& point)
+	{
+
+		thePoints.push_back(point);
+		for (unsigned int i = 0; i < numberOfDimensions; ++i)
+			theDimensions.at(i).push_back(point[i]);
+		theIds.push_back(point.getId());
+
+	
+	}
+
+	void push_back(FKDPoint<TYPE, numberOfDimensions> && point)
```

Comment:
Can we have also `emplace_back()` as a syntactic sugar to avoid repeating the explicit type? E.g. minimally along
```
template <typename ... Args>
void emplace_back(Args&&... arg) {
   push_back(FKDPoint<TYPE, numberOfDimensions>(std::forward<Args>(args)...);
}
```